GUI Boilerplate and Computer Case GUI Beginnings

rschulman opened the pull request:
This draft PR puts in place a few GUI necessities and adds an almost complete case GUI. It does raise some questions about whether we want to make the different case tiers different classes entirely or continue to make them just variables within one class.

rschulman committed:
Create some fundamentals for gui management and partial completion of Computer GUI

rschulman committed:
Merge conflicts from upstream

lucsoft commented:
I would like to keep it as variable as there is not really a change between tthe caseses expact the slots

lucsoft commented on src/main/kotlin/li/cli/oc/Components.kt:
```diff
@@ -33,10 +39,10 @@ object Components {
-        CaseCreative("casecreative", Case(4)),
+        CaseOne("case1", Case(1, FabricBlockSettings.of(Material.METAL).nonOpaque())),
+        CaseTwo("case2", Case(2, FabricBlockSettings.of(Material.METAL).nonOpaque())),
+        CaseThree("case3", Case(3, FabricBlockSettings.of(Material.METAL).nonOpaque())),
```
Why was this changed? The block settings are the same on the Cases

rschulman replied:
I'll be honest, I don't precisely remember. I think it came about from needing to access the `settings` in Case.kt before they were passed on to TecBlock? I'd need to do a bit more digging to be sure.

lucsoft replied:
in Kotlin there is a init function so you dont need to pass it

rschulman replied:
Yeah, I have no idea why I thought I needed this. It works fine the other way. I'll push a fix.

lucsoft replied:
cool

rschulman committed:
Revert changes to `Case` constructor in `Case.kt` and `Case` creation in `Components.kt`

rschulman committed:
Propagate case tier into Case BlockEntity and ScreenHandler

rschulman committed:
Add new texture for slots

Slots are not working properly yet because the `CaseScreenHandler` is seemingly created twice per GUI opening. Once from `CaseEntity` and once from its registration in`Components.kt`. The latter is generic across Case tiers, and so right now passes in 0 for the tier, which causes an array index out of bounds error.

lucsoft commented:
Remeber to update this pr i made big changes (every tecblock is now an blockwithentity)

rschulman commented:
I'm working on updating it now, but the new `CaseEntity` doesn't extend `BlockEntity` directly, so I'm not sure how to override its functions. Right now IDEA is telling me that `createMenu` (e.g.) overrides nothing. Do you know how I should proceed?

rschulman committed:
Rebase, plus move CaseEntity functions into the new Case.kt file

rschulman committed:
Change Case into an ExtendedScreenHandler, trying to get correct case tier to affect slot layout

rschulman committed:
Add toTag and fromTag for Case entity to store and recover entity tier

rschulman committed:
Add and align slots for tier 1 and 2, and slot textures for tier 1

omixann commented:
Hm, slots seem to have scaling issues. I tried to fix it with no luck.

lucsoft commented:
I look into it after im done with my work related stuff

omixann commented:
I'm currently copying code from [upstream](https://github.com/MightyPirates/OpenComputers/blob/master-MC1.7.10/src/main/scala/li/cil/oc/client/gui/DynamicGuiContainer.scala#L146), let's see if that works!

omixann commented:
![image](https://user-images.githubusercontent.com/17623703/103654785-9e897880-4f66-11eb-8265-66e282e5a7fb.png)
@rschulman expect a PR soon

rschulman commented:
Awesome, thanks @jmi2k!

rschulman commented:
@jmi2k  Just checking in on this. I don't think I've seen a PR from you, unless I'm missing something!

lucsoft commented:
Its in https://github.com/jmi2k/OpenComputers/tree/fix-ui
but keep in mind my focus is on OC2

omixann commented:
@rschulman can't you see it [here](https://github.com/rschulman/OpenComputers/pull/1)? Did I do something wrong? It's the first time I send PR to different repos from the same fork, maybe I messed up something.
@lucsoft what do you mean by OC2?

lucsoft commented:
@jmi2k as in the readme i had teamed up with the Creator of OpenComputers to help bring OpenComputers 2 for Forge & Fabric

omixann commented:
Wait there *is* a v2 of OC in the works! That's amazing!

lucsoft commented:
> Wait there _is_ a v2 of OC in the works! That's amazing!

Yes its now a Linux Computer that has a lua runtime
